stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -243,10 +247,112 @@ swift::_matchMetadataByMangledTypeName(const llvm::StringRef typeName,
+  // archetypes matches the number of child nodes in the demangled type.
+  //
+  // TODO: support secondary archetypes, where NumParams != NumPrimaryParams;
+  // not sure if there is enough information available at runtime to do this.
+  auto &genericParams = ntd->GenericParams;
+  if (!genericParams.hasGenericParams() ||
+      genericParams.NumPrimaryParams != typeListNode->getNumChildren() ||
+      genericParams.NumParams != genericParams.NumPrimaryParams)
+    return nullptr;
+
+  auto templateName = mangleNode(templateTypeNode); // XXX cache this
+  if (ntd->Name.get() != templateName)
+    return nullptr;
+
+  auto pattern = ntd->getGenericMetadataPattern();
+  void **arguments = new void *[pattern->NumKeyArguments];
+
+  for (size_t i = 0, witnessIndex = 0;
+       i < genericParams.NumPrimaryParams; i++) {
+    auto param = genericParams.getParameterAt(i);
+    auto metadata = findMetadataForNode(typeListNode->getChild(i));
+    if (metadata == nullptr) {
+      // unknown type
+      delete arguments;
+      return nullptr;
+    }
+
+    for (size_t j = 0; j < param->NumWitnessTables; j++, witnessIndex++) {
+      auto &protocol = param->Protocols[j];
+      auto wt = swift_conformsToProtocol(metadata, protocol);
+      if (wt == nullptr) {
+        // does not conform to associated type requirement, bail
+        delete arguments;
+        return nullptr;
+      }
+
+      arguments[genericParams.NumPrimaryParams + witnessIndex] =
+        const_cast<void *>(reinterpret_cast<const void *>(wt));
+    }
+
+    arguments[i] = const_cast<void *>(reinterpret_cast<const void *>(metadata));
+  }
+
+  auto metadata = swift_getGenericMetadata(pattern, arguments);
+
+  delete arguments;
+  return metadata;
+}
+
 static const Metadata *
-_searchTypeMetadataRecords(const TypeMetadataState &T,
-                           const llvm::StringRef typeName) {
+matchMetadataNode(const llvm::StringRef typeNodeName,
+                  const NodePointer typeNode,
+                  const Metadata *metadata,
+                  const NominalTypeDescriptor *ntd) {
+  if (typeNode->getKind() != Node::Kind::Type || !typeNode->hasChildren())
+    return nullptr;
+
+  auto childNode = typeNode->getChild(0);
+  const Metadata *foundMetadata = nullptr;
+
+  switch (childNode->getKind()) {
+  case Node::Kind::BoundGenericClass:
+  case Node::Kind::BoundGenericEnum:
+  case Node::Kind::BoundGenericStructure:
+    if (ntd != nullptr)
+      foundMetadata = matchMetadataGenericNode(typeNodeName, childNode,
+                                               metadata, ntd);
+    break;
+  default:
+    foundMetadata = matchMetadataTypeNode(typeNodeName, childNode,
+                                          metadata, ntd);
+    break;
+  }
+
+  return foundMetadata;
+}
+
+// caller must acquire lock
+static const Metadata *
+iterateTypeMetadata(
+  const llvm::StringRef nameRef,
+  const NodePointer node,
+  llvm::function_ref<const Metadata *(llvm::StringRef,
```

Comment:
This would be more readable if you used an alias, like

    using MetadataSearchFn = const Metadata *(llvm::StringRef,
                                          const NodePointer,
                                          const Metadata *,
                                         const NominalTypeDescriptor *);